Fix buggy email extraction with AUTH verb

diegosalvi opened the pull request:
This issue is needed to cover a bug introduced with #18.

The issue arise when a https://tools.ietf.org/html/rfc4954 compliant client connect to subethamail. Mail commands will be received in this form:
MAIL FROM:<john+@example.org> AUTH=<>

See https://tools.ietf.org/html/rfc4954#section-5

This patch change extraction behavior checking for brackets open/close and not expanding email address to the last '>'

Added some testcases

diegosalvi commented:
Seen in a production environment

```
20-12-23-16-22-21       SMTP connection from xxxxxx/xx.xx.xx.xx, new connection count: 2
20-12-23-16-22-21       Server: 220 xxxxxx ESMTP xxxxx
20-12-23-16-22-21       Client: EHLO xxxxx
20-12-23-16-22-21       Server: 250-xxxxxxxx^M
250-8BITMIME^M
250-STARTTLS^M
250-CHUNKING^M
250-AUTH PLAIN LOGIN^M
250 Ok
20-12-23-16-22-21       Client: STARTTLS
20-12-23-16-22-21       Server: 220 Ready to start TLS
20-12-23-16-22-21       Server: 250 Ok
20-12-23-16-22-22       Client: EHLO xxxxx
20-12-23-16-22-22       Server: 250-xxxxxxxx^M
250-8BITMIME^M
250-STARTTLS^M
250-CHUNKING^M
250-AUTH PLAIN LOGIN^M
250 Ok
20-12-23-16-22-22       Client: AUTH PLAIN Axxxxxxxxxxxxxxxo
20-12-23-16-22-22       Server: 235 Authentication successful.
20-12-23-16-22-22       Client: MAIL FROM:<xxxxxxxx> AUTH=<>
20-12-23-16-22-22       Server: 553 <xxxxxxxx> AUTH=<> Invalid email address.
20-12-23-16-22-22       Client: RSET
```

davidmoten reviewed (commented):
On a quick glance I'm surprised not to see regex used for this extraction. Why didn't you?

diegosalvi commented:
Cause matching open/close brackets can be really error prone with regex.
Perhaps with 3 regex...
* match no brackets: `\s*(?<address>\S+).*`
* match rfc brackets: `\s*[<]{1}(?<address>[^ >]+)\s*[>]{1}.*`
* match syntax for #18 (personal name): `\s*[<]{1}[^<]+(?<address>[^ >]+)\s*[>]{1}\s*[>]{1}.*`

Hopefully there shouldn't be more than two <> (I presume...)
We could check the 'personal', if doesn't match check the rfc than the no brackets (rfc match the personal name too but in a wrong way, no brackets too matches the other two in a wrong way)

Perhaps the no brackets could be rewritten as `\s*(?<address>[^<]*\S+).*`

davidmoten commented:
Ok not to use regex. Does your new code have full test coverage?

diegosalvi commented:
Mmm the personal name pattern is wrong.. it should contains the personal name too

diegosalvi commented:
I've added coverage for the AUTH verb with null sender <> that broke old parsing. There aren't any more cases (at least for auth) that could mislead the address recongnition.

Still... I've added some more tests with null sender that were missing

davidmoten commented:
What does the standard say about how the server should respond when the syntax is wrong? For example, no closing `>`?

diegosalvi commented:
> [RFC5321 Section 3.3](https://tools.ietf.org/html/rfc5321#section-3.3):
>
>       If the mailbox specification is not acceptable for
>       some reason, the server MUST return a reply indicating whether the
>       failure is permanent (i.e., will occur again if the client tries to
>       send the same address again) or temporary (i.e., the address might be
>       accepted if the client tries again later).

> [RFC3463 Section 3.2](https://tools.ietf.org/html/rfc3463#section-3.2)
>
>       X.1.7   Bad sender's mailbox address syntax
> 
>       The sender's address was syntactically invalid.  This can apply
>       to any field in the address.

> [IANA Simple Mail Transfer Protocol (SMTP) Enhanced Status Codes Registry](https://www.iana.org/assignments/smtp-enhanced-status-codes/smtp-enhanced-status-codes.xhtml)
>
>       X.1.7 | Bad sender's mailbox address syntax | Not given | The sender's address was syntactically invalid. This can apply to any field in the address.


I'd say it should return a `501 Bad sender address syntax` o a `501 5.1.7 Bad sender address syntax` to support extended codes.

Currently we return a `553 <emailAddress> Invalid email address.`
I didn't want to change such behavior too.

davidmoten commented:
Looks good, thanks for the contribution @diegosalvi!

davidmoten commented:
version 5.2.8 has been released to Maven Central with your fix.
